getusermedia.html

```diff
@@ -3479,6 +3479,10 @@ <h2>Methods</h2>
                         inside of <a>"video"</a> and video-only constraints inside of
                         <a>"audio"</a> are simply ignored rather than causing
                         <code>OverconstrainedError</code>.</li>
+                        <li><p>If <var>CS</var> contains a constrainable property
+                        that is <a>required</a> and whose name is not in the
+                        list of <a>allowed required constraints for device selection</a>,
+                        jump to the step labeled <em>Constraint Failure</em> below.</p></li>
```

Comment:
We have a [precedent](https://github.com/w3c/mediacapture-image/pull/261#issuecomment-698514049) in screen-capture we can copy for navigating these dictionaries.

Also, for consistency, we think in https://github.com/w3c/mediacapture-image/pull/261#issuecomment-698516937 it should throw `TypeError`.

```suggestion
                        <li><p>If <var>CS</var> contains a member whose name is not in the
                        list of <a>allowed required constraints for device selection</a>,
                        and whose value in turn is a dictionary containing a member named
                        <code>min</code>, <code>max</code>, or <code>exact</code>,
                        then <a>reject</a> <var>p</var> with a {{TypeError}}, and abort
                        these steps.</p></li>
```

Reply:
> We have a [precedent](https://github.com/w3c/mediacapture-image/pull/261#issuecomment-698514049) in screen-capture we can copy for navigating these dictionaries.

What is the benefit? I prefer referring to a concept (required, defined in https://w3c.github.io/mediacapture-main/#dfn-required-constraints) instead of trying to inline the definition of the concept. I would tend to fix media capture-screenshare so that it refers to the concept in mediacapture-main.

Or do you want to disallow these constraints in advanced constraints too?

I am ok with TypeError instead of OverConstrained.

Reply:
@youennf All inputs can be validated using the terminology of the incoming data type. No need to involve concepts.

The input we're reading here is a dictionary, which does not contain a "constrainable property" (which is even the wrong concept, it should at least be "constraint") but a [member](https://heycam.github.io/webidl/#dfn-dictionary-member), which has a name (aka key) and a value.

The problem I have using our concepts is that they're complicated and harder to navigate than the WebIDL primitives.

For example, "constrainable property" != "constraint". "[required](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-required)" != "[required constraint](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-required-constraints)".

Specifically, I worry that the sentence _"If CS contains a constrainable property that is **required**"_ is vague about whether to reject on `advanced` constraints which are also [required](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-required) and also something the CS may "contain".

I also worry implementers may miss that the dictionary to inspect is nested one level down.

Reply:
> I worry that the sentence _"If CS contains a constrainable property that is **required**"_ is vague about whether to reject on `advanced` constraints which are also [required](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-required) and also something the CS may "contain".

That is why the term "required" is a link that points to the following definition:
"required constraints are all non-advanced constraints that are required."

We could rephrase a bit the sentence though to use the term "required constraint".
This term is already used in getUserMedia just below. I do not think implementors will miss the difference here.

I think this PR is clear about allowing advanced constraints that are required.
We might want to further restrict this but this could be done as a follow-up.

Reply:
> That is why the term "required" is a link that points to the following definition:
"required constraints are all non-advanced constraints that are [required](https://w3c.github.io/mediacapture-main/#dfn-required)."

@youennf but that's not where it links. You're using `<a>required</a>` which links to the [required](https://w3c.github.io/mediacapture-main/#dfn-required) definition which doesn't preclude `advanced` constraints. This is the kind of confusion I'm talking about.

 Also the terminology is wrong. How about:
```suggestion
                        <li><p>If <var>CS</var> contains a member that is a
                        <a>required constraint</a> and whose name is not in the
                        list of <a>allowed required constraints for device selection</a>,
                        then <a>reject</a> <var>p</var> with a {{TypeError}}, and abort
                        these steps.</p></li>
```

Reply:
I see, thxs for the update.